### PR TITLE
Refactor the C++ code

### PR DESCRIPTION
**Pull request by csukuangfj**
The goal is to
- [x] Refactor the code to make it more readable, maintainable, and extensible
- [ ] Add more comments/documentation
- [ ]  Add CI tests

### CODE REVIEW

File: sherpa/csrc/offline-transducer-modified-beam-search-decoder.h [hunk 1]
```diff
@@ -16,7 +16,7 @@ class OfflineTransducerModifiedBeamSearchDecoder
  public:
   explicit OfflineTransducerModifiedBeamSearchDecoder(
       OfflineTransducerModel *model, int32_t num_active_paths_)
```
Comment (pkufool) [inline sherpa/csrc/offline-transducer-modified-beam-search-decoder.h:18]:
```suggestion
      OfflineTransducerModel *model, int32_t num_active_paths)
```

File: sherpa/csrc/offline-transducer-modified-beam-search-decoder.h [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+#include "sherpa/csrc/offline-transducer-decoder.h"
+#include "sherpa/csrc/offline-transducer-model.h"
+
+namespace sherpa {
+
+class OfflineTransducerModifiedBeamSearchDecoder
+    : public OfflineTransducerDecoder {
+ public:
+  explicit OfflineTransducerModifiedBeamSearchDecoder(
```
Comment (pkufool) [inline sherpa/csrc/offline-transducer-modified-beam-search-decoder.h:17]: Do we need `explicit` here?

File: sherpa/cpp_api/offline-recognizer.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
+// sherpa/cpp_api/offline-recognizer.h
+//
```
Comment (csukuangfj) [inline sherpa/cpp_api/offline-recognizer.h:2]: This file can also be reused for CTC decoding.

**Comment by csukuangfj** [comment]
This PR is getting large.  I have refactored the C++ code and left the code related to Python unchanged.
Will refactor the Python code and update the doc in a separate PR. 

Will merge this PR once all CI tests pass.

**Pull request merged**